tensorflow/c/c_api_experimental.cc

```diff
@@ -576,6 +578,60 @@ void TF_MakeInternalErrorStatus(TF_Status* status, const char* errMsg) {
   status->status = tensorflow::errors::Internal(errMsg);
 }
 
+struct TF_CheckpointReader : public tensorflow::checkpoint::CheckpointReader {
+  using tensorflow::checkpoint::CheckpointReader::CheckpointReader;
+  std::vector<std::string> variable_list;
+};
+
+TF_CheckpointReader* TF_NewCheckpointReader(const char* filename, TF_Status* status) {
+  TF_CheckpointReader* reader = new TF_CheckpointReader(filename, status);
+  if (!status->status.ok()) return nullptr;
+  auto m = reader->GetVariableToDataTypeMap();
+  for( auto it = m.begin(); it != m.end(); ++it)
+    reader->variable_list.push_back(it->first);
+  std::sort(reader->variable_list.begin(), reader->variable_list.end());
+  return reader;
+}
+
+void TF_DeleteCheckpointReader(TF_CheckpointReader* reader) { delete reader; }
+
+int TF_CheckpointReaderHasTensor(TF_CheckpointReader* reader, const char* name) {
+  return reader->HasTensor(name);
+}
+
+const char* TF_CheckpointReaderGetVariable(TF_CheckpointReader* reader, int index) {
+  return reader->variable_list[index].c_str();
+}
+
+int TF_CheckpointReaderSize(TF_CheckpointReader* reader) {
+  return reader->variable_list.size();
+}
+
+TF_DataType TF_CheckpointReaderGetVariableDataType(TF_CheckpointReader* reader, const char* name) {
+  auto m = reader->GetVariableToDataTypeMap();
+  return static_cast<TF_DataType>(m[name]);
+}
+
+TF_Tensor* TF_CheckpointReaderGetTensor(TF_CheckpointReader* reader, const char* name, TF_Status* status) {
+  std::unique_ptr<tensorflow::Tensor> tensor;
+  reader->GetTensor(name, &tensor, status);
+  if (!status->status.ok()) return nullptr;
+  return tensorflow::TF_TensorFromTensor(*tensor.get(), status);
+}
+
+void TF_CheckpointReaderGetVariableShape(TF_CheckpointReader* reader, const char* name, int64_t* dims, int num_dims) {
+  auto m = reader->GetVariableToShapeMap();
+  for (int i=0; i<num_dims; i++){
+    dims[i] = m[name].dim_size(i);
```

Comment:
Hang on to m[name]? Not that there'll be that many dims.

Do we also want to validate num_dims?

Reply:
I try to follow some api design in other c code, I think the `num_dims` should be get from the `TF_CheckpointReaderGetVariableNumDims`?

Reply:
Right, that's the expected usage. A couple of the others claim to except if the wrong value is passed in.

Reply:
Then do we need to also pass a `TF_Status` for validating the `num_dims`?

Reply:
Sounds reasonable to me, for consistency with https://github.com/tensorflow/tensorflow/blob/4d56e9709f1864699be7241a3d1938816d9f31c5/tensorflow/c/c_api.cc#L1172

Reply:
added